src/dbnode/storage/bootstrap/bootstrapper/peers/source.go

```diff
@@ -642,73 +648,123 @@ func (s *peersSource) readIndex(
 		zap.Int("concurrency", concurrency),
 	)
 
+	groupFn := bootstrapper.NewShardTimeRangesTimeWindowGroups
+	groupedByBlockSize := groupFn(shardsTimeRanges, dataBlockSize)
+
 	workers := xsync.NewWorkerPool(concurrency)
 	workers.Init()
-	for shard, ranges := range shardsTimeRanges {
-		shard, ranges := shard, ranges
+	for _, group := range groupedByBlockSize {
+		shardsTimeRanges := group.Ranges
 		wg.Add(1)
+		// NB(bodu): This is fetching the data for all shards for a block of time.
 		workers.Go(func() {
-			defer wg.Done()
-
-			iter := ranges.Iter()
-			for iter.Next() {
-				target := iter.Value()
-				size := dataBlockSize
-				for blockStart := target.Start; blockStart.Before(target.End); blockStart = blockStart.Add(size) {
-					currRange := xtime.Range{
-						Start: blockStart,
-						End:   blockStart.Add(size),
-					}
-
-					// Helper lightweight lambda that should get inlined
-					var markedAnyUnfulfilled bool
-					markUnfulfilled := func(err error) {
-						markedAnyUnfulfilled = true
-						s.markIndexResultErrorAsUnfulfilled(r, resultLock, err,
-							shard, currRange)
-					}
+			var (
+				docsPool = s.opts.DocumentArrayPool()
+				batch    = docsPool.Get()
+			)
+			defer func() {
+				docsPool.Put(batch)
+			}()
+
+			for shard, ranges := range shardsTimeRanges {
+				iter := ranges.Iter()
+				for iter.Next() {
+					target := iter.Value()
+					size := dataBlockSize
+					for blockStart := target.Start; blockStart.Before(target.End); blockStart = blockStart.Add(size) {
+						currRange := xtime.Range{
+							Start: blockStart,
+							End:   blockStart.Add(size),
+						}
 
-					metadata, err := session.FetchBootstrapBlocksMetadataFromPeers(ns.ID(),
-						shard, currRange.Start, currRange.End, resultOpts)
-					if err != nil {
-						// Make this period unfulfilled
-						markUnfulfilled(err)
-						continue
-					}
-					for metadata.Next() {
-						_, dataBlock := metadata.Current()
+						// Helper lightweight lambda that should get inlined
+						var markedAnyUnfulfilled bool
+						markUnfulfilled := func(err error) {
+							markedAnyUnfulfilled = true
+							s.markIndexResultErrorAsUnfulfilled(r, resultLock, err,
+								shard, currRange)
+						}
 
-						inserted, err := s.readBlockMetadataAndIndex(r, resultLock, dataBlock,
-							idxOpts, resultOpts)
+						indexBlockDocumentsBuilder, err := r.IndexResults().GetOrAddDocumentsBuilder(
+							blockStart,
+							idxOpts,
+							resultOpts,
+						)
 						if err != nil {
 							// Make this period unfulfilled
 							markUnfulfilled(err)
+							continue
 						}
+						flushBatch := bootstrapper.CreateFlushBatchFn(resultLock, batch, indexBlockDocumentsBuilder)
 
-						if !inserted {
-							// If the metadata wasn't inserted we finalize the metadata.
-							dataBlock.Finalize()
+						metadata, err := session.FetchBootstrapBlocksMetadataFromPeers(ns.ID(),
```

Comment:
So, I think for the peers indexing read side we can just index from whatever we just persisted to disk.